opentracing-kotlin/src/main/kotlin/org/zalando/opentracing/kotlin/TracerExtension.kt

```diff
@@ -0,0 +1,94 @@
+package org.zalando.opentracing.kotlin
+
+import io.opentracing.Scope
+import io.opentracing.Span
+import io.opentracing.SpanContext
+import io.opentracing.Tracer
+import io.opentracing.propagation.Format.Builtin.TEXT_MAP
+import io.opentracing.propagation.TextMapAdapter
+import io.opentracing.tag.Tags
+
+inline fun <R> Tracer.trace(span: Span, block: (Span) -> R): R {
+
+    var scope: Scope? = null
+    var closed = false
+
+    try {
+        scope = scopeManager().activate(span)
+
+        return block(span)
+    } catch (e: Exception) {
+        span.record(e)
+        try {
+            scope?.close()
+            closed = true
+        } catch (closeException: Exception) {
+            // Do Nothing
+        }
+        throw e
+    } finally {
+        try {
+            if (!closed)
+                scope?.close()
+            span.finish()
+        } catch (closeException: Exception) {
+            // Do Nothing
+        }
+    }
+}
+
+inline fun <R> Tracer.trace(
+    operation: String,
+    spanBuilder: Tracer.SpanBuilder.() -> Unit,
+    block: (Span) -> R
+): R {
+    val builder = buildSpan(operation)
+    builder.apply(spanBuilder)
+    val span = builder.start()
+    return trace(span, block)
+}
+
+inline fun <R> Tracer.trace(
+    operation: String,
+    component: String? = null,
+    kind: String? = null,
+    parent: SpanContext? = null,
+    block: (Span) -> R
+): R {
+    return trace(
+        operation = operation,
+        spanBuilder = {
+            parent?.let { asChildOf(parent) }
+            component?.let { withTag(Tags.COMPONENT, component) }
+            kind?.let { withTag(Tags.SPAN_KIND, kind) }
+        },
+        block = block
+    )
+}
+
+fun Span.record(ex: java.lang.Exception) {
+    log(
+        mapOf(
+            "event" to Tags.ERROR.key,
+            "error.object" to ex
+        )
+    )
+    setTag(Tags.ERROR, true)
+}
+
+fun Tracer.getSpanContextAsMap(span: Span? = activeSpan()) =
```

Comment:
One way to make this more in line with the `extract` method from below could be:

```kotlin
fun Tracer.inject(span: Span? = activeSpan(), map: Map<String, String> = mutableMapOf()) {
    span?.let {
        inject(it.context, TEXT_MAP, TextMapAdapter(map))
    }
}
```

Reply:
Might even be handy in cases where the caller already has a mutable Map at hand for some reason.

Reply:
1. param type must be MutableMap
2. map has to be returned

However - this seems to be overkill, desiginng for edge cases - where you can just use the plain tracing API

Reply:
@piyushmor please specify the return types of all methods explicitly - makes it easier to understand the code

Reply:
Then what about `Tracer.injectIntoMap`?

Reply:
I like that - I think it's a good idea to keep the `inject` and `extract` names from the original API.

Reply:
I don't feel comfortable in adding a mutable map here as those are rare in kotlin. the only reason we return a mutable map here is because the opentracing api forces it in that fashion. I would even like to hide the fact that we are returning a mutable map if possible(as done by specifying the return type explicitly). The caller can still merge multiple maps if needed after calling this method but I don't think we will come across a mutable map in kotlin usually.

Reply:
> I don't feel comfortable in adding a mutable map here as those are rare in kotlin. the only reason we return a mutable map here is because the opentracing api forces it in that fashion. I would even like to hide the fact that we are returning a mutable map if possible(as done by specifying the return type explicitly). The caller can still merge multiple maps if needed after calling this method but I don't think we will come across a mutable map in kotlin usually.

yes, I agree with this, and this is another reason not to add the parameter.